### PR TITLE
Add unittests for ServiceNowTicket

### PR DESCRIPTION
**Pull request by pzubaty**
After merging this, only test on production is missing. There is small issue of fixing compatibility with main repo. Merge should be done before adding add_attachment.

### CODE REVIEW

**Comment by pzubaty** [comment]
@dranck @rpavelka please review

File: tests/test_servicenow.py
Comment (pzubaty) [inline tests/test_servicenow.py]: I am thinking about adding some graph of inheritance or something.

Comment (pzubaty) [reply tests/test_servicenow.py]: done with the last commit (see TestServiceNowTicket)

File: ticketutil/servicenow.py
Comment (rpavelka) [inline ticketutil/servicenow.py]: reason for commented code leaved there?

File: ticketutil/servicenow.py
Comment (rpavelka) [inline ticketutil/servicenow.py]: would recommend different name with URL in it. Also consider that CamelCase is recommended by PEP8 for classes and functions should be all lowercase with underscored: https://www.python.org/dev/peps/pep-0008/#function-names

File: ticketutil/ticket.py
Comment (rpavelka) [inline ticketutil/ticket.py:115]: why there is no raise to raise RequestException again?

Also e.args may be empty.

File: ticketutil/ticket.py
Comment (rpavelka) [inline ticketutil/ticket.py:108]: why to do not `self.session = s` here instead?

File: ticketutil/ticket.py
Comment (pzubaty) [reply ticketutil/ticket.py:115]: Well I have copied whole module from the master repo, so that's something @dranck should be asked.

File: ticketutil/ticket.py
Comment (pzubaty) [reply ticketutil/ticket.py:108]: this should be directed to Danny (see prev. comment)

File: ticketutil/servicenow.py
Comment (pzubaty) [reply ticketutil/servicenow.py]: I copied this line from jira.py, but during unit tests I realized that's not going to work. I can remove it I guess.

File: ticketutil/ticket.py
Comment (rpavelka) [reply ticketutil/ticket.py:115]: OK, I did asked @dmranck by mail... I see I should add you there too. I will ask Danny to respond here...

Comment (dmranck) [reply ticketutil/ticket.py:115]: Good point, @rpavelka :) I'm working on better exception handling and should be pushing an update in the next couple days. I'll open a PR for it and include you and @pzubaty.

So, for this specific case, I raise the exception in __init__() lines 45-47. If a session object is not returned from _create_requests_session(), then I raise a TicketException:

```
        self.s = self._create_requests_session()
        if not self.s:
            raise TicketException("Error authenticating to {0}.".format(self.auth_url))
```
I'm also adding some new checks in __init__() to verify the project and optional ticket_id being passed in actually exist. Because if those don't exist, a Ticket object should not be created and an exception should be raised.

File: ticketutil/ticket.py
Comment (dmranck) [reply ticketutil/ticket.py:108]: self.s is being set in __init__() on line 45. This allows me to raise a TicketException if a session was not successfully created in _create_requests_session(). Does this make sense? Or do you think there's a better way to do it? Any suggestions are welcome. :)

File: ticketutil/ticket.py
Comment (rpavelka) [reply ticketutil/ticket.py:115]: I see. So if this is inline with your intentions, respects code in other places and you plan to change it anyway, then I am fine with that.

File: ticketutil/ticket.py
Comment (rpavelka) [reply ticketutil/ticket.py:108]: I would propose creating attribute here instead of returning it. But maybe I am missing some use case. Then I would propose raising error in try block immediately after cleanup. Instead:
```python
        try:
            r = s.get(self.auth_url)
            r.raise_for_status()
            logging.debug("Create requests session: Status Code: {0}".format(r.status_code))
            logging.info("Successfully authenticated to {0}".format(self.ticketing_tool))
            return s
        except requests.RequestException as e:
            logging.error("Error authenticating to {0}.".format(self.auth_url))
            logging.error(e.args[0])
            s.close()
```
I would propose:
```python
        try:
            r = s.get(self.auth_url)
            r.raise_for_status()
        except requests.RequestException as e:
            logging.error("Error authenticating to {0}.".format(self.auth_url))
            if e.args:
                logging.error(e.args[0])
            s.close()
            raise

        logging.debug("Create requests session: Status Code: {0}".format(r.status_code))
        logging.info("Successfully authenticated to {0}".format(self.ticketing_tool))
        self.s = s
```
and to remove `self.s = ` in `__init__`. I prefer to keep logical gears in single place whenever possible. But I understand the nice thing that you see explicitly new attribute being created in `__init__`. IMHO to reraise immediately is better. But I see returning new session and adding it as an attribute in `__init__` has its merits too.

But again maybe I am missing some reason for this...

Comment (dmranck) [reply ticketutil/ticket.py:108]: I think the reason I did it this way was executing the "raise" in your proposal above resulted in a kind of messy stack trace of the following exceptions:

requests.exception.ConnectionError, requests.packages.urllib3.exceptions.MaxRetryError, requests.packages.urllib3.exceptions.NewConnectionError, ConnectionRefusedError.

By raising the TicketException, we could pass in our own message and the traceback was cleaner:

ticketutil.ticket.TicketException: Error authenticating to http://0.0.0.0:300054/projects.json.

With the new code where I'm verifying the project and ticket_id upon creation of a Ticket object, I do like how the exceptions are cleanly raised in __init__():

```python
self.s = self._create_requests_session()
if not self.s:
    raise TicketException("Error authenticating to {0}.".format(self.auth_url))

if not self._verify_project(self.project):
    raise TicketException("Project {0} is not valid.".format(self.project))

if self.ticket_id and not self._verify_ticket_id(self.ticket_id):
    raise TicketException("Ticket {0} is not valid.".format(self.ticket_id))
```
But then again, these exceptions could be immediately raised inside those methods as well, like you're proposing... And I do see your point about how raising it immediately is better. I'll open a PR soon with my code containing the verify_project / verify_ticket_id updates and we can continue the discussion there after you guys take a look.

**Comment by dmranck** [comment]
@pzubaty, I pulled your code down but am having trouble running the unittests. How are you currently running them?

edit: I should have been more specific. I'm trying to run through the command line, but am seeing:
`ImportError: No module named 'ticketutil'`
when servicenow.py is trying to run:
`from ticketutil.ticket import Ticket, LOG_LEVEL`

**Comment by pzubaty** [comment]
I was running it from main ticketutil directory with `python3 -m unittest ./tests/*.py`
EDIT: Actually it should have been `python3 -m unittests tests.test_servicenow`

**Comment by pzubaty** [comment]
@dmranck I would like to merge this branch and open a pull request for the main repo, so it can be used. Is there anything else or can I do that? :)

File: ticketutil/servicenow.py
Comment (dmranck) [inline ticketutil/servicenow.py]: You don't need to import LOG_LEVEL here

File: ticketutil/servicenow.py
Comment (dmranck) [inline ticketutil/servicenow.py]: Lets hold off on all the "raise"s in the servicenow.py methods for now.

The next big thing on my to do list for ticketutil is to return data from each method that includes if the request was successful. I'd like to go this route instead of raising the Request Exception each time one of the ticket methods fail.

See line 85, 183, 207, 244, 262, 291, 315, 344

File: ticketutil/servicenow.py
Comment (dmranck) [inline ticketutil/servicenow.py:366]: Where is this used? I don't see this method being called from anywhere.

File: tests/test_servicenow.py
Comment (dmranck) [inline tests/test_servicenow.py:9]: Is there a better way we can handle the imports from the tests/ directory? This is a little messy. I can look more into this as well.

File: tests/test_servicenow.py
Comment (dmranck) [inline tests/test_servicenow.py]: should be "full-length"

**Review by dmranck — CHANGES_REQUESTED** [review]
Unit tests look great! I think you can merge this branch into your fork's repo after taking a look at my comments.

Before merging to the main repo, I'd like to see the following two methods added:  _verify_project() and _verify_ticket_id(). (Let me know if you need help). I recently added these checks across the other four tools in the following commit, and servicenow will need them as well to work with ticket.py: https://github.com/dmranck/ticketutil/commit/2fbc3b766f0c3a5272f2952c77e5d6fe8f537530 . You code should also be rebased because there have been a lot of changes on master.

As for add_attachment(), I don't think that would hold up the merge to the main repo, but I'd like to see that added soon if possible. add_attachment() for servicenow is a task I could take on, so that I get more familiar with servicenow. :)

File: ticketutil/servicenow.py
Comment (pzubaty) [reply ticketutil/servicenow.py:366]: It's a helper function which creates URL for DevOps One application.[1] I didn't feel it would be good idea to include it in the class itself, but it's related. I can add an example to /examples 
[1] https://redhat.service-now.com/pnt

File: tests/test_servicenow.py
Comment (pzubaty) [reply tests/test_servicenow.py:9]: It's still less messy than the alternative. You can see this line in E2E-Container unit tests. Without it there are problems when patch calls servicenow module.

File: ticketutil/servicenow.py
Comment (pzubaty) [reply ticketutil/servicenow.py]: I see it now, thx.

**Review by dmranck — APPROVED** [review]
Thanks for changing the 'raise' to 'return False'. I know it's not ideal at the current moment for the unit tests, but I'll be working on returning better data very soon (starting this week). 

I think everything in this branch looks good and ready to merge.

File: ticketutil/servicenow.py
Comment (dmranck) [reply ticketutil/servicenow.py:366]: One more quick question on this - Should this helper function really be in ticketutil? Or should it be in your external code that calls ticketutil? Would anyone else find this function useful, or is it really just for your specific use case of ServiceNowTicket? 

Just trying to think about this from a general perspective - I don't want any Red Hat Internal functions in ticketutil if they wouldn't benefit anyone outside of Red Hat.

**Pull request merged**
